### ISSUE TITLE
Optimize Speaker Diarization for Apple Silicon using Metal

### ISSUE THREAD

**Issue by gabrielanyosa1** [gabrielanyosa1/AudioTranscriptionApp#1]
# Metal Optimization for Speaker Diarization

## Current Status
Currently, the speaker diarization process runs on CPU only, not taking full advantage of Apple Silicon's Neural Engine and Metal capabilities. This results in suboptimal performance and higher resource usage than necessary on M1/M2/M3 Macs.

## Objectives
1. Implement Metal backend support for PyTorch operations in the diarization pipeline
2. Optimize pyannote.audio pipeline for Metal
3. Utilize Apple Neural Engine where possible
4. Maintain compatibility with CPU fallback for non-Apple Silicon systems

### Research and Implementation
- Investigate Metal Performance Shaders (MPS) backend compatibility with pyannote.audio
- Profile current CPU performance to establish baseline metrics
- Modify model loading to prefer Metal devices when available
- Test and optimize memory usage patterns with Metal
- Implement efficient CPU-Metal memory transfers

### Code Modifications Needed
1. Update PyTorch configuration:
```python
import torch
def get_optimal_device():
    if torch.backends.mps.is_available():
        return torch.device('mps')
    return torch.device('cpu')
```

2. Modify pipeline initialization:
```python
device = get_optimal_device()
pipeline = Pipeline.from_pretrained(
    "pyannote/speaker-diarization",
    use_auth_token=os.getenv('HUGGINGFACE_TOKEN')
).to(device)
```

3. Optimize tensor operations:
```python
# Current CPU-only implementation
def _align_transcription_with_speakers(self, segments, diarization):
    aligned = []
    for segment in segments:
        # CPU processing
        ...
    return aligned

# Proposed Metal-optimized implementation
def _align_transcription_with_speakers(self, segments, diarization):
    device = get_optimal_device()
    aligned = []
    # Batch process segments using Metal when available
    if device.type == 'mps':
        # Optimized batch processing
        ...
    else:
        # Fallback CPU processing
        ...
    return aligned
```

### Phase 1: Foundation
1. Add Metal device detection and selection
2. Implement basic Metal pipeline loading
3. Add CPU fallback mechanism
4. Create performance logging infrastructure

### Phase 2: Optimization
1. Profile memory transfer patterns
2. Optimize batch processing for Metal
3. Implement memory usage improvements
4. Add thermal monitoring

### Phase 3: Testing & Validation
1. Benchmark performance across different file sizes
2. Validate speaker detection accuracy
3. Test CPU fallback scenarios
4. Document performance improvements

### Performance Improvements
- Significant reduction in processing time
- Lower CPU utilization
- Better thermal performance
- Improved battery efficiency on laptops

### Technical Benefits
- Better utilization of Apple Silicon capabilities
- More efficient memory usage
- Improved scalability for larger files
- Maintained or improved accuracy

### Performance Targets
- 2-3x speedup in diarization processing
- 30-40% reduction in CPU usage
- 25% reduction in memory footprint
- No degradation in speaker detection accuracy

### Validation Methods
- Benchmark suite for processing times
- CPU/Memory usage profiling
- Thermal monitoring data
- Accuracy comparison with CPU baseline

## Technical References
- [PyTorch MPS Documentation](https://pytorch.org/docs/stable/notes/mps.html)
- [Apple Metal Performance Shaders](https://developer.apple.com/metal/pytorch/)
- [PyTorch Memory Management](https://pytorch.org/docs/stable/notes/cuda.html#memory-management)
- [pyannote.audio Documentation](https://github.com/pyannote/pyannote-audio)

## Additional Notes
- Implementation should maintain compatibility with CPU processing for non-Apple Silicon systems
- Consider adding configuration options for users to force CPU processing if needed
- Document any known limitations or edge cases with Metal processing
